sdk/src/client/api/block_builder/transaction_builder/mod.rs

```diff
@@ -40,14 +43,132 @@ use crate::{
     },
 };
 
-/// Working state for the input selection algorithm.
+impl Client {
+    /// Builds a transaction using the given inputs, outputs, addresses, and options.
+    pub async fn build_transaction(
+        &self,
+        addresses: impl IntoIterator<Item = Address>,
+        outputs: impl IntoIterator<Item = Output>,
+        options: TransactionOptions,
+    ) -> crate::client::Result<PreparedTransactionData> {
+        let outputs = outputs.into_iter().collect::<Vec<_>>();
+        let addresses = addresses.into_iter().collect::<Vec<_>>();
+        let protocol_parameters = self.get_protocol_parameters().await?;
+        let creation_slot = self.get_slot_index().await?;
+
+        let slot_commitment_id = self.get_issuance().await?.latest_commitment.id();
+        let reference_mana_cost = if let Some(issuer_id) = options.issuer_id {
+            Some(self.get_account_congestion(&issuer_id, None).await?.reference_mana_cost)
+        } else {
+            None
+        };
+        let remainder_address = match options.remainder_value_strategy {
+            RemainderValueStrategy::ReuseAddress => None,
+            RemainderValueStrategy::CustomAddress(address) => Some(address),
+        };
+
+        let hrp = protocol_parameters.bech32_hrp();
+
+        let mut available_input_ids = HashSet::new();
+        for address in &addresses {
+            let mut cursor = None;
+            loop {
```

Comment:
This loop and cursor stuff isn't needed, that's already handled inside of `output_ids()`

Reply:
oh nice!